File: Lab 1.2/Lab 1.2 on Python/funcs.py
import pickle
from datetime import time
from os.path import exists
import os


CONST_PATH = "records.dat"
dayStart = time(9,0)
dayEnd = time(17,0)
addInfoAboutFinish = False
timePeriodsList = []

class TimePeriod :
    def __init__(self, tStart, tEnd):
        self.tStart = tStart
        self.tEnd = tEnd


def ClearOrContinue():
    if (exists(CONST_PATH)):
        print("records.bat already exists. Clear it (1) or continue populating (2) ?   (1/2) ")
        ans = int(input())
        print()
        if (ans==1):
            os.remove(CONST_PATH)
        else:
            global addInfoAboutFinish
            addInfoAboutFinish = True
            AddExistingRecordsToList()



def AddExistingRecordsToList():
    with open(CONST_PATH, 'rb') as file:
        while (True):
            try:
                surname = pickle.load(file)
            except Exception:
                break

            tStart = pickle.load(file)
            # tStart = time(int(stStart[0:2]), int(stStart[3:]))

            tEnd = pickle.load(file)
            # tEnd = time(int(stEnd[0:2]), int(stEnd[3:]))

            timePeriodsList.append(TimePeriod(tStart,tEnd))

def PrintRecordings():
    print("All records :")
    print()
    with open(CONST_PATH, 'rb') as file:
        while (True):
            try:
                surname = pickle.load(file)
            except Exception:
                break
            tStart = pickle.load(file)
            tEnd = pickle.load(file)

            print("Surname: " + surname)
            print("Service start time: " + str(tStart.strftime("%H:%M")))
            print("Service end time: " + str(tEnd.strftime("%H:%M")))
            print()


def MakeRecord(surname, tStart, tEnd ):
    with open(CONST_PATH,'ab') as file:
        pickle.dump(surname,file)
        pickle.dump(tStart, file)
        pickle.dump(tEnd, file)

def Populate() :
    while (True):
        overlaping = False
        global  addInfoAboutFinish
        if(addInfoAboutFinish):
            print("Enter a surname of your client (type \"finish\" to finish) :")
        else:
            print("Enter a surname of your client :")
        surname = input()
        if (surname=="finish"):
            break
        print()

        print ("Enter a customer arrival time (in format hh:mm) :")
        stStart = input()
        try:
            tStart = time(int(stStart[0:2]), int(stStart[3:]))
        except Exception:
            print()
            print("Incorrect time format. Record could not be added!")
            print()
            continue
        print()

        print ("Enter a customer service end time (in format hh:mm) :")
        stEnd = input()
        try:
            tEnd = time(int(stEnd[0:2]), int(stEnd[3:]))
        except Exception:
            print()
            print("Incorrect time format. Record could not be added!")
            print()
            continue
        print()

        if (tEnd<=tStart):
            print("Start time cannot be greater then or equal to end time! Record could not be added!")
            print()
            continue

        if (tStart<dayStart or tEnd>dayEnd):
            print("Working day starts at 09:00 and ends at 17:00. Record could not be added! ")
            print()
            continue

        for span in timePeriodsList:
            if (tStart< span.tEnd and span.tStart < tEnd):
                print("This time period overlaps the existing one. Record could not be added!")
                overlaping = True
                break

        if (overlaping):
            continue

        timePeriodsList.append(TimePeriod(tStart, tEnd))
        MakeRecord(surname, tStart, tEnd)
        addInfoAboutFinish = True

def FindAndPrintGaps():
    print()
    print("Time when manager was free :")
    print()
    sortedList = timePeriodsList
    sortedList.sort(key= lambda TimePeriod: TimePeriod.tStart)
    freeTime = False
    count = 0
    position = dayStart
    for period in sortedList:
        if (position < period.tStart):
            count+=1
            print(str(position.strftime("%H:%M")) + " - " + str(period.tStart.strftime("%H:%M")))
        position = period.tEnd

    if (sortedList[-1].tEnd < dayEnd):
        count+=1
        print(str(sortedList[-1].tEnd.strftime("%H:%M")) + " - " + str(dayEnd.strftime("%H:%M")))

    print()
    print("A total amout of gaps : " + str(count))

File: Lab 1.2/Lab 1.2 on CS/Funcs.cs
using System;
namespace Lab
{
    public partial class Program
    {
        static List<TimePeriod> timePeriodsList = new List<TimePeriod>();
        const string path = "records.dat";
        static TimeOnly dayStart = new TimeOnly(9, 0);
        static TimeOnly dayEnd = new TimeOnly(17, 0);
        static bool addInfoAboutFinish = false;

        struct TimePeriod
        {
            public TimeOnly tStart;
            public TimeOnly tEnd;
        }


        // FULL PROCESS OF PUPULATING BINARY FILE
        private static void Populate()
        {

            ConsoleKeyInfo key;
            bool overlaping = false;
            BinaryWriter writer = new BinaryWriter(File.Open(path, FileMode.Append));

            while (true)
            {
                TimeOnly tStart;
                TimeOnly tEnd;
                overlaping = false;

                if (addInfoAboutFinish)
                {
                    Console.WriteLine(@"Enter a surname of your client (press ""Home"" button to finish) :");
                }
                else
                {
                    Console.WriteLine("Enter a surname of your client :");
                }
                key = Console.ReadKey();
                if (key.Key == ConsoleKey.Home)
                {
                    break;
                }

                string surname = key.KeyChar.ToString() + Console.ReadLine();
                Console.WriteLine();


                Console.WriteLine("Enter a customer arrival time (in format hh:mm) :");
                string? startTime = Console.ReadLine();
                try
                {
                    tStart = new TimeOnly(Convert.ToInt32(startTime[0..2]), Convert.ToInt32(startTime[3..5]));
                }
                catch (Exception)
                {
                    Console.WriteLine();
                    Console.WriteLine("Incorrect time format. Record could not be added!");
                    Console.WriteLine();
                    continue;
                }
                Console.WriteLine();



                Console.WriteLine("Enter a customer service end time (in format hh:mm) :");
                string? endTime = Console.ReadLine();
                try
                {
                    tEnd = new TimeOnly(Convert.ToInt32(endTime[0..2]), Convert.ToInt32(endTime[3..5]));
                }
                catch (Exception)
                {
                    Console.WriteLine();
                    Console.WriteLine("Incorrect time format. Record could not be added!");
                    Console.WriteLine();
                    continue;
                }
                Console.WriteLine();



                if (tEnd <= tStart)
                {
                    Console.WriteLine("Start time cannot be greater then or equal to end time! Record could not be added!");
                    Console.WriteLine();
                    continue;
                }

                if (tStart < dayStart || tEnd > dayEnd)
                {
                    Console.WriteLine("Working day starts at 09:00 and ends at 17:00. Record could not be added! ");
                    Console.WriteLine();
                    continue;
                }

                foreach (var span in timePeriodsList)
                {
                    if (tStart < span.tEnd && span.tStart < tEnd)
                    {
                        Console.WriteLine("This time period overlaps the existing one. Record could not be added!");
                        Console.WriteLine();
                        overlaping = true;
                        break;
                    }

                }

                if (overlaping)
                {
                    continue;
                }

                timePeriodsList.Add(new TimePeriod { tEnd = tEnd, tStart = tStart });
                MakeRecord(surname, tStart, tEnd, writer);
                addInfoAboutFinish = true;

            }
            writer.Close();

        }


        // ASK USER (CLEAR FIlE OR CONTINUE) IF FILE ALREADY EXIST
        private static void ClearOrContinue()
        {
            if (File.Exists(path))
            {
                Console.WriteLine("records.bat already exists. Clear it (1) or continue populating (2) ?   (1/2) ");
                int ans = Convert.ToInt32(Console.ReadLine());
                Console.WriteLine();
                if (ans == 1)
                {
                    File.Delete(path);
                }
                else
                {
                    addInfoAboutFinish = true;
                    AddExistingRecordsToList();
                }
                Console.Clear();
            }
        }

        // WRITE A RECORD IN A BINARY FILE
        private static void MakeRecord(string surname, TimeOnly tStart, TimeOnly tEnd, BinaryWriter writer)
        {

            writer.Write(surname);
            writer.Write(tStart.ToString("HH:mm"));
            writer.Write(tEnd.ToString("HH:mm"));

        }

        private static void AddExistingRecordsToList()
        {
            using (BinaryReader reader = new BinaryReader(File.Open("records.dat", FileMode.Open)))
            {

                while (reader.PeekChar() > -1)
                {
                    string surname = reader.ReadString();
                    string startTime = reader.ReadString();
                    string endTime = reader.ReadString();

                    TimeOnly tEnd = new TimeOnly(Convert.ToInt32(endTime[0..2]), Convert.ToInt32(endTime[3..5]));
                    TimeOnly tStart = new TimeOnly(Convert.ToInt32(startTime[0..2]), Convert.ToInt32(startTime[3..5]));

                    timePeriodsList.Add(new TimePeriod { tEnd = tEnd, tStart = tStart });

                }
            }
        }


        // PRINT ALL RECORDINGS FROM A BINARY FILE IN CONSOLE
        private static void PrintRecordings()
        {
            Console.Clear();
            Console.WriteLine("All records :");
            Console.WriteLine();
            using (BinaryReader reader = new BinaryReader(File.Open("records.dat", FileMode.Open)))
            {
                // пока не достигнут конец файла
                // считываем каждое значение из файла
                while (reader.PeekChar() > -1)
                {
                    string surname = reader.ReadString();
                    string startTime = reader.ReadString();
                    string endTime = reader.ReadString();
                    //TimeOnly tEnd = new TimeOnly(Convert.ToInt32(endTime[0..2]), Convert.ToInt32(endTime[3..5]));
                    //TimeOnly tStart = new TimeOnly(Convert.ToInt32(startTime[0..2]), Convert.ToInt32(startTime[3..5]));
                    //timePeriodsList.Add(new TimePeriod { tEnd = tEnd, tStart = tStart });
                    Console.WriteLine("Surname: " + surname);
                    Console.WriteLine("Service start time: " + startTime);
                    Console.WriteLine("Service end time: " + endTime);
                    Console.WriteLine();

                }
            }
        }

        private static void FindAndPrintGaps(List<TimePeriod> timePeriodsList)
        {
            Console.WriteLine();
            Console.WriteLine("Time when manager was free :");
            Console.WriteLine();
            List<TimePeriod> sorted = timePeriodsList.OrderBy(x => x.tStart).ToList();
            bool freeTime = false;
            int count = 0;
            TimeOnly position = dayStart;
            foreach (var period in sorted)
            {
                if (position < period.tStart)
                {
                    count++;
                    Console.WriteLine(position + " - " + period.tStart);
                }
                position = period.tEnd;

            }
            if (sorted.Last().tEnd < dayEnd)
            {
                count++;
                Console.WriteLine(sorted.Last().tEnd + " - " + dayEnd);
            }
            Console.WriteLine();
            Console.WriteLine("A total amout of gaps : " + count);
        }
    }
}